Implement actor sketch

mikayla-maki opened the pull request:
Fixes https://github.com/penumbra-zone/tower-actor/issues/1

TODO:

- [x] Figure out good tests for each of the error conditions
- [x] Double check that exports work correctly.
- [x] Get a pass from the penumbra dev team
- [x] Document the library!

OUT OF DATE COMMENTARY:

Modeling the errors in this crate proved to be a bit trickier than I expected. The original code uses `anyhow` + `Box<dyn Error + Send + Sync + 'static>` to erase all error information but this didn't feel like the right approach for a library. I ended up going with [sled's](http://sled.rs/errors.html) approach of returning a `Result<Result<S, E>, ActorError>` to capture the difference between unrecoverable queue errors and theoretically recoverable consumer errors. This also required me to make the actual ActorFuture itself an enum, to capture the case where the queue was closed between `poll_ready()` and `call()`. Curious to hear what y'all think :)

mikayla-maki committed:
Implement actor sketch

mikayla-maki committed:
Added tests, discovered that one test condition is impossible according to the contract of PollSender, refactored code to simplify interfaces

mikayla-maki commented:
The error condition referenced above, which required the actor future to become an enum, turned out to be impossible. The documentation of [`PollSender::close()`](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.PollSender.html#method.close) guarantees that you can still send on a queue that has been dropped / `.closed()` / panicked until all of it's permits have been used up. As such, I was able to simplify the error handling significantly and retain the design as shown in the sketch. I added a couple tests for some common cases that I could think of, honestly there's not that much actual code being executed here. Most of the effort has been in setting up the types / interfaces.

mikayla-maki committed:
Add compile error guard for projects compiled without tokio unstable

mikayla-maki commented:
Added some compile flags to ensure that projects which include tower_actor but don't use `tokio_unstable` get a useful error message. Also, made sure that all of the re-exports look nice.

mikayla-maki committed:
Disable doc tests

hdevalence commented:
On the error handling front, the reason to prefer a boxed error rather than nesting multiple concrete error type is that it's the only practical way to achieve composability: if you have a single Result<Rsp, BoxError>, the caller code doesn't need to know the details of the whole layer stack, but they can still use downcasting to handle specific error types if they want. By contrast, nested error types reify the layer stack in the response type, so that changing it later (eg rearranging the order of layers) breaks caller code.

hdevalence commented on src/lib.rs:
```diff
@@ -1,14 +1,179 @@
-    left + right
+pub use tracing;
+
+#[cfg(not(tokio_unstable))]
+mod tower_actor {
+    compile_error!("Tower actor must be compiled with a `--cfg tokio_unstable` rustc flag");
+}
+
+pub use tower_actor::*;
+
+#[cfg(tokio_unstable)]
+mod tower_actor {
+    pub(crate) mod actor;
+    pub(crate) mod future;
+    pub(crate) mod message;
+
+    pub use actor::*;
+    pub use future::*;
+    pub use message::*;
```
The only tokio_unstable feature used here is the named task used to spawn the actor, right?

I think it'd be cleaner to make the crate compile unconditionally, and limit the cfg-gating to the spawn call, using the cfg to switch between either calling spawn directly and ignoring the task name, or using the unstable Builder API if it's available. (Maybe we'll get an unused variable warning in the unavailable case, I think that doesn't matter too much)

hdevalence commented on src/tower_actor/actor.rs:
```diff
@@ -0,0 +1,96 @@
+        // before calling `call()`, we can safely proceed without checking that the queue isn't closed.
+        debug_assert!(!self.queue.is_closed());
+
+        let span = req.create_span();
```
I think this is a holdover from the pd code, which hooked custom span creation into the service definitions. But in a generic context I don't think it's appropriate for the service to create custom spans as well as doing the actor part — we should just use Span::current() here, and if the caller wants to do custom spam handling, they should do that by composing in an additional tower layer. That way, we only do one thing, and let the user compose behavior as desired.

hdevalence reviewed (changes requested):
This looks great so far, thanks so much! I only had two minor comments.

mikayla-maki committed:
Remove custom span creation code
Make crate compilation unconditional, and only expose the named() API if tokio unstable is compiled

mikayla-maki commented:
Changes implemented!

mikayla-maki committed:
Remove vestigal return types from actor constructor callback

mikayla-maki commented:
While reviewing the code, I also noticed that the type of the actor constructor callback was a little funny, in the sketch it returns a `Result<(), MessageErrorType>`, but the join handle carrying that return is immediately dropped in the constructors. I've switched it to returning the unit type to more clearly express this. 

That said, I think it could be cool if the Actor was itself a future that could be `.await`to get at that return type. But to do that I think we'd need to stuff some types in somewhere to express 'what is this actor's result and error' and then we'd probably have to make an 'actor reference' that holds the poll sender and can be cloned around. Not too hard, but a big change from the existing API. Also I should probably go have a weekend at some point 😅

hdevalence commented:
Ah, yeah. I think the writeup I did was unclear. The idea behind having the actor future return an error is so that at some point, we can extend the implementation to have a shared "error slot" (like in the Buffer implementation) like Arc<Mutex<Option<BoxError>>> or something, then

- wrap the actor future in a shim that checks if it exited with an error and sets the error slot
- hook the poll_ready implementation to check the error slot and fail the Service if the Actor failed (IIRC there may be some complexity here about needing to clone the error? maybe it has to be Option<Arc<BoxError>> or something so that multiple Services can share the internal error?)

I think we want the future the user passed in to resolve to a Result, even if we currently discard the error, so that later, we can improve the implementation to handle errors better without having to change the external API.

For now I think it's fine to discard the error as an implementation shortcoming, and plan to fix it up later.

mikayla-maki commented:
Ahhh, I see. I did read the note about that but I didn’t quite put all the pieces together. I’ll look into the buffer stuff and maybe add an error slot implementation later. Thanks for the context :D

mikayla-maki committed:
Return result types to new() functions

mikayla-maki commented:
Looked up the buffer shim implementation, and it seemed like more effort than this Sunday is calling for. Restored the results as part of a planned future expansion and hopefully this should be good to go now :)

hdevalence committed:
correct cfg-gating for tokio_unstable

hdevalence committed:
add crate metadata to Cargo.toml
